executor/src/witgen/identity_processor.rs

```diff
@@ -48,8 +48,12 @@ impl<'a, 'b, T: FieldElement> Machines<'a, 'b, T> {
         self.machines.len()
     }
 
-    pub fn iter_mut(&'b mut self) -> impl Iterator<Item = &'b mut KnownMachine<'a, T>> {
-        self.machines.iter_mut().map(|m| &mut **m)
+    pub fn iter_mut<'c>(&'c mut self) -> impl Iterator<Item = &'c mut KnownMachine<'a, T>> {
+        self.machines
+            .iter_mut()
+            .map(|m| &mut **m)
+            .collect::<Vec<_>>()
```

Comment:
Why do we need to re-collect this into a vector?

Reply:
It has to do with lifetimes. Analogous to what happens in `split()`, we want to re-borrow the `&'b mut` into a `&'c mut` where `'c` is shorter than `'b`. If I remove the `collect()` I get a compiler error:

```
error[E0700]: hidden type for `impl Iterator<Item = &'c mut KnownMachine<'a, T>>` captures lifetime that does not appear in bounds
  --> executor/src/witgen/identity_processor.rs:52:9
   |
23 | impl<'a, 'b, T: FieldElement> Machines<'a, 'b, T> {
   |          -- hidden type `std::iter::Map<std::slice::IterMut<'c, &'b mut KnownMachine<'a, T>>, [closure@executor/src/witgen/identity_processor.rs:52:38: 52:41]>` captures the lifetime `'b` as defined here
...
51 |     pub fn iter_mut<'c>(&'c mut self) -> impl Iterator<Item = &'c mut KnownMachine<'a, T>> {
   |                                          ------------------------------------------------- opaque type defined here
52 |         self.machines.iter_mut().map(|m| &mut **m)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0700`.
```

I think the reason we can't return reference of lifetime `'b` anymore is because we now call `iter_mut()` inside `Generator` (to finalize secondary VMs). One way around that would be to move that up to `executor/src/witgen/mod.rs` (which already before used `iter_mut()` to call `take_witness_col_values()` on all machines). One could even have `take_witness_col_values()` receive a `mutable_state: &mut MutableState<'a, '_, T, Q>` and do the finalization in there.

Do you think that would be better?

Reply:
> One could even have `take_witness_col_values()` receive a `mutable_state: &mut MutableState<'a, '_, T, Q>` and do the finalization in there.

I did this in bea1cea9d76dbc27bc2c0e297d26aa1c41199060. So this change has been undone!